### DISCUSSION TITLE
Universal2 wheel for macOS

### DISCUSSION THREAD

**Discussion by kuldeepsingh0694** [python-pillow/Pillow#6385] [Q&A] [answered]
Does Pillow have a Universal2 whl for any version , im building a python GUI app which keeps failing in pyinstaller build due to incompatible Architecture. Apparently all the other libs im using have universal2 whl (pandas,matplotlib,numpy).  

System: macOS 12.4 , Apple Silicon (M1 pro)

**Comment by radarhere** [comment]
Hi. Do you really need universal2 wheels, or will arm64 wheels suffice? We provide macOS arm64 wheels that should work for you on your M1. You can see at https://pypi.org/project/Pillow/#files that we have
- Pillow-9.1.1-cp310-cp310-macosx_11_0_arm64.whl
- Pillow-9.1.1-cp39-cp39-macosx_11_0_arm64.whl
- Pillow-9.1.1-cp38-cp38-macosx_11_0_arm64.whl

I would have thought that numpy does not produce universal2 wheels, as I don't see them at https://pypi.org/project/numpy/#files

**Reply by kuldeepsingh0694** [reply]
Hi, Thanks for the reply, yes i need a universal2 wheel as I'm packing is app to work on both Apple Silicon and Intel Chips for macOS.
Currently pyinstaller cant use arm64 whl to build a universal2 app (Incompatibile Binary Arch Error : PIL is not a fat binary (pillow vesion9.1.1 arm64))

PS: I think numpy stopped producing universal2 for newer build. For that reason I'm using numpy 1.21.6 where universal2 whl is available 

and does using lipo with pillow arm64 and [x86_64.whl ] can produce the fat binary?

**Reply by hugovk** [reply]
If you're fine using an old version of Pillow in addition to old NumPy, Pillow 9.0.1 had a universal2 wheel:

* Pillow-9.0.1-cp310-cp310-macosx_10_10_universal2.whl 
* https://pypi.org/project/Pillow/9.0.1/#files

**Reply by kuldeepsingh0694** [reply]
Hi ,i have checked this universal2.whl but there are some issue with it (Image import fails on macOS 10.13 with Pillow 9.1.0 #6179)
it throw importerror dlopen () symbol not found flat namespace

**Reply by kuldeepsingh0694** [reply]
PyDev console: starting.
Python 3.10.5 (v3.10.5:f377153967, Jun  6 2022, 12:36:10) [Clang 13.0.0 (clang-1300.0.29.30)] on darwin
import PIL
import PIL
from PIL import Image
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/code.py", line 90, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
  File "/Applications/PyCharm.app/Contents/plugins/python/helpers/pydev/_pydev_bundle/pydev_import_hook.py", line 21, in do_import
    module = self._system_import(name, *args, **kwargs)
  File "/Users/kdsin/PycharmProjects/DesktopAppUniversal2/venv/lib/python3.10/site-packages/PIL/Image.py", line 89, in <module>
    from . import _imaging as core
  File "/Applications/PyCharm.app/Contents/plugins/python/helpers/pydev/_pydev_bundle/pydev_import_hook.py", line 21, in do_import
    module = self._system_import(name, *args, **kwargs)
ImportError: dlopen(/Users/kdsin/PycharmProjects/DesktopAppUniversal2/venv/lib/python3.10/site-packages/PIL/_imaging.cpython-310-darwin.so, 0x0002): symbol not found in flat namespace '_jpeg_resync_to_restart'
PIL.__version__
'9.0.1'

platform.platform()
'macOS-12.4-arm64-arm-64bit'

platform.architecture()
('64bit', '')



this is a new virtualenv , installed the above whl with pip install path.whl , python is 3.10.5 universal2 installer

**Comment by radarhere** [comment, marked as answer]
How about this? [Pillow-9.1.1-cp310-cp310-macosx_universal2.whl.zip](https://github.com/python-pillow/Pillow/files/8957051/Pillow-9.1.1-cp310-cp310-macosx_universal2.whl.zip)

I created it using [delocate](https://pypi.org/project/delocate/).

```python
from delocate import fuse

in1 = "Pillow-9.1.1-cp310-cp310-macosx_10_10_x86_64.whl"
in2 = "Pillow-9.1.1-cp310-cp310-macosx_11_0_arm64.whl"
out = "Pillow-9.1.1-cp310-cp310-macosx_universal2.whl"

fuse.fuse_wheels(in1, in2, out)
```

**Reply by kuldeepsingh0694** [reply]
(venv) kdsin@bcd074488ea4 DesktopAppUniversal2 % pip install /Users/kdsin/Downloads/Pillow-9.1.1-cp310-cp310-macosx_universal2.whl    
ERROR: Pillow-9.1.1-cp310-cp310-macosx_universal2.whl is not a supported wheel on this platform.

platform.platform()
'macOS-12.4-arm64-arm-64bit'

platform.architecture()
('64bit', '')

 python is 3.10.5 universal2 installer
macOS 12.4

**Reply by radarhere** [reply]
Wait, I think I just had the wrong filename. How about this? [Pillow-9.1.1-cp310-cp310-macosx_10_10_universal2.whl.zip](https://github.com/python-pillow/Pillow/files/8957232/Pillow-9.1.1-cp310-cp310-macosx_10_10_universal2.whl.zip)

**Reply by kuldeepsingh0694** [reply]
can install now, tell me check few things with it thanks a lot!
